### ISSUE TITLE
ImapClient.Idle is not cancelled

### ISSUE THREAD

**Issue by ekalchev** [jstedfast/MailKit#1025]
**Describe the bug**
We observe that sometimes ImapClient.Idle is not cancelled when doneToken and cancellationToken are cancelled. We first cancel doneToken and then cancellationToken as stated in the documentation.

It is not reproduced every time. We noticed that this happens when computer is waking up from sleep and ImapClient was in idle when computer went to sleep. Probably during the computer sleep server dropped the socket because of timeout on the server side, I think this might be related.

**Platform (please complete the following information):**
 - Windows 10
 - .NET Framework 4.8
 - MailKit Version - checkout at https://github.com/jstedfast/MailKit/commit/adc0aecb1a2aff67418cbd38cf968a5593cb4828

*Additional information*

ImapClient.Idle call stack after tokens cancellation

`ntdll.dll!00007ffb6abdc0f4()	Unknown
mswsock.dll!00007ffb672782ed()	Unknown
mswsock.dll!00007ffb67278032()	Unknown
ws2_32.dll!00007ffb6aa01920()	Unknown
System.ni.dll!00007ffb38be12b2()	Unknown
[Managed to Native Transition]	
System.dll!System.Net.Sockets.Socket.Receive(byte[] buffer, int offset, int size, System.Net.Sockets.SocketFlags socketFlags, out System.Net.Sockets.SocketError errorCode)	Unknown
MailKit.dll!MailKit.Net.NetworkStream.Read(byte[] buffer, int offset, int count)	Unknown
System.dll!System.Net.FixedSizeReader.ReadPacket(byte[] buffer, int offset, int count)	Unknown
System.dll!System.Net.Security._SslStream.StartFrameHeader(byte[] buffer, int offset, int count, System.Net.AsyncProtocolRequest asyncRequest)	Unknown
System.dll!System.Net.Security._SslStream.StartReading(byte[] buffer, int offset, int count, System.Net.AsyncProtocolRequest asyncRequest)	Unknown
System.dll!System.Net.Security._SslStream.ProcessRead(byte[] buffer, int offset, int count, System.Net.AsyncProtocolRequest asyncRequest)	Unknown
System.dll!System.Net.Security.SslStream.Read(byte[] buffer, int offset, int count)	Unknown
MailKit.dll!MailKit.Net.Imap.ImapStream.ReadAheadAsync(int atleast, bool doAsync, System.Threading.CancellationToken cancellationToken)	Unknown
mscorlib.dll!System.Runtime.CompilerServices.AsyncTaskMethodBuilder<int>.Start<MailKit.Net.Imap.ImapStream.<ReadAheadAsync>d__59>(ref MailKit.Net.Imap.ImapStream.<ReadAheadAsync>d__59 stateMachine)	Unknown
MailKit.dll!MailKit.Net.Imap.ImapStream.ReadAheadAsync(int atleast, bool doAsync, System.Threading.CancellationToken cancellationToken)	Unknown
MailKit.dll!MailKit.Net.Imap.ImapStream.ReadTokenAsync(string specials, bool doAsync, System.Threading.CancellationToken cancellationToken)	Unknown
mscorlib.dll!System.Runtime.CompilerServices.AsyncTaskMethodBuilder<System.__Canon>.Start<MailKit.Net.Imap.ImapStream.<ReadTokenAsync>d__74>(ref MailKit.Net.Imap.ImapStream.<ReadTokenAsync>d__74 stateMachine)	Unknown
MailKit.dll!MailKit.Net.Imap.ImapStream.ReadTokenAsync(string specials, bool doAsync, System.Threading.CancellationToken cancellationToken)	Unknown
MailKit.dll!MailKit.Net.Imap.ImapCommand.StepAsync(bool doAsync)	Unknown
mscorlib.dll!System.Runtime.CompilerServices.AsyncTaskMethodBuilder<bool>.Start<MailKit.Net.Imap.ImapCommand.<StepAsync>d__83>(ref MailKit.Net.Imap.ImapCommand.<StepAsync>d__83 stateMachine)	Unknown
MailKit.dll!MailKit.Net.Imap.ImapCommand.StepAsync(bool doAsync)	Unknown
MailKit.dll!MailKit.Net.Imap.ImapEngine.IterateAsync(bool doAsync)	Unknown
mscorlib.dll!System.Runtime.CompilerServices.AsyncTaskMethodBuilder.Start<MailKit.Net.Imap.ImapEngine.<IterateAsync>d__170>(ref MailKit.Net.Imap.ImapEngine.<IterateAsync>d__170 stateMachine)	Unknown
MailKit.dll!MailKit.Net.Imap.ImapEngine.IterateAsync(bool doAsync)	Unknown
MailKit.dll!MailKit.Net.Imap.ImapEngine.RunAsync(MailKit.Net.Imap.ImapCommand ic, bool doAsync)	Unknown
mscorlib.dll!System.Runtime.CompilerServices.AsyncTaskMethodBuilder.Start<MailKit.Net.Imap.ImapEngine.<RunAsync>d__171>(ref MailKit.Net.Imap.ImapEngine.<RunAsync>d__171 stateMachine)	Unknown
MailKit.dll!MailKit.Net.Imap.ImapEngine.RunAsync(MailKit.Net.Imap.ImapCommand ic, bool doAsync)	Unknown
MailKit.dll!MailKit.Net.Imap.ImapClient.IdleAsync(System.Threading.CancellationToken doneToken, bool doAsync, System.Threading.CancellationToken cancellationToken)	Unknown
mscorlib.dll!System.Runtime.CompilerServices.AsyncTaskMethodBuilder.Start<MailKit.Net.Imap.ImapClient.<IdleAsync>d__99>(ref MailKit.Net.Imap.ImapClient.<IdleAsync>d__99 stateMachine)	Unknown
MailKit.dll!MailKit.Net.Imap.ImapClient.IdleAsync(System.Threading.CancellationToken doneToken, bool doAsync, System.Threading.CancellationToken cancellationToken)	Unknown
MailKit.dll!MailKit.Net.Imap.ImapClient.Idle(System.Threading.CancellationToken doneToken, System.Threading.CancellationToken cancellationToken)	Unknown`

**Comment by jstedfast** [comment]
Could you try upgrading to the latest? There's been a number of fixes to this logic.

**Comment by ekalchev** [comment]
I already did, I'll keep monitoring and will update the issue if I have more info.

**Comment by jstedfast** [comment]
We can leave it open if you've confirmed that it still exists in 2.6.

**Comment by ekalchev** [comment]
This is a problem in v2.6 too. It happens randomly on different users but only when the computer wake ups from sleep.

**Comment by jstedfast** [comment]
I don't think this is fixable. The problem is that the synchronous Stream API doesn't allow for cancellation.

My recommendation would be to switch to the async API.

**Comment by ekalchev** [comment]
I switched from sync version to imapClient.IdleAsync().GetAwaiter().GetResult() and I still get that issue. Any other ideas?

**Comment by jstedfast** [comment]
The stack trace *should* be different in this case. Can you try posting an updated stack trace?

I have a feeling that if using the async API doesn't already work, there's probably nothing I can do to make it work, but an updated stack trace might reveal something that I'm not considering.

The good news is that the .NET Core networking team is working on improving .NET Core's network support. They asked me to review some new network connection-related API proposals a few months ago and they looked promising. Through my discussion with one of the developers on that team, I also learned that .NET 5 will add async cancellation to more socket APIs.

I guess what I'm saying is that even if this is not fixable *now*, it might be fixable once .NET 5 comes out.

FWIW, back when I was working on writing [Evolution](https://wiki.gnome.org/Apps/Evolution/) in the early 2000's, we struggled with the same issue of sockets becoming hung when resuming a laptop from sleep mode. Part of the problem at the time (and may be the same issue you are hitting now) was that the thread that was trying to abort the network call would end up hanging as well because calling close() on a hung socket would block until the socket freed itself.

Actually... hmmm 🤔 

That gives me an idea...

Take a look at https://github.com/jstedfast/MailKit/blob/master/MailKit/Net/NetworkStream.cs#L179 - notice that if the socket is still connected (which it probably thinks it is in your scenario), MailKit calls `Socket.Shutdown (Both)`.

In .NET, there are multiple ways to close a socket connection:

- [Close](https://docs.microsoft.com/en-us/dotnet/api/system.net.sockets.socket.close?view=netcore-3.1)
- [Disconnect](https://docs.microsoft.com/en-us/dotnet/api/system.net.sockets.socket.disconnect?view=netcore-3.1)
- [DisconnectAsync](https://docs.microsoft.com/en-us/dotnet/api/system.net.sockets.socket.disconnectasync?view=netcore-3.1)
- [Dispose](https://docs.microsoft.com/en-us/dotnet/api/system.net.sockets.socket.dispose?view=netcore-3.1)
- [Shutdown](https://docs.microsoft.com/en-us/dotnet/api/system.net.sockets.socket.shutdown?view=netcore-3.1)

The problem with `Shutdown(Both)` is that it requests that the socket send a packet to the remote host. If you are trying to resume a laptop after being suspended long enough for the other end to have already dropped the connection, then it will be impossible for the socket to send those packets to the remote host and so it will just hang waiting for a response (until the request times out).

Unfortunately, having read the docs before, it's hard to know exactly what each of these various methods does when you call them.

Would you be willing to experiment a bit with the various methods above to see if there is one that works in your scenario?

In the meantime, maybe I can reach out to the .NET Core network team and see if they have any suggestions for me.

**Comment by ekalchev** [comment]
Unfortunately I don't have call stack but I'll try to find one.

**Comment by jstedfast** [comment]
Just for clarification - are you cancelling the `doneToken` or the `cancellationToken`?

When you "cancel" the `doneToken`, it attempts to send a `"DONE\r\b"` command to the server which could also be the culprit. I just want to make sure I'm chasing the right boogyman ;-)

**Comment by ekalchev** [comment]
I am cancelling both.

**Comment by ekalchev** [comment]
Thanks. I'll try them and will let you know the outcome.

**Comment by jstedfast** [comment]
In the patches above, I switched from using Shutdown() to Disconnect(). I just got a response from the .NET Core team which suggested I use Close() instead of Shutdown() and they confirmed my suspicions about what Shutdown(Both) was doing.

If the above changes don't work for you, we'll try Close() instead of Disconnect().

**Comment by ekalchev** [comment]
I see an issue with this [change](https://github.com/jstedfast/MailKit/commit/83f09887dbd1f8c737742a46c49d85c514ff612a)

When I cancel both done and cancellation token the thread that is blocked by

imapClient.IdleAsync(done, cancellationToken).GetAwaiter().GetResult() is not released for about 30-40sec. At first I thought it was a deadlock but inspecting the Tasks wait chain shows that awaiters in the mailkit are waiting for a task that is scheduled but not executed yet. It takes about 30-40 sec to complete. Here is the actual line of code that is awaiters are waiting so they can continue.

![image](https://user-images.githubusercontent.com/10009149/88035119-54dcdb00-cb4a-11ea-8e05-1ca3aae79411.png)

Reverting this change above fixed the problem.

**Comment by jstedfast** [comment]
If I revert the commit that sends the DONE command asynchronously, that fixes just this 40 second hang? Or is your main issue also fixed?

I'll revert the async DONE change either way if that improves things for you.

**Comment by ekalchev** [comment]
That fixes the 40 seconds hand. The main issue is hard to reproduce and it will takes a week of monitoring until someone complain about it.

**Comment by jstedfast** [comment]
Ok, thanks - I just reverted the async DONE command.

**Comment by ekalchev** [comment]
[Here](https://stackoverflow.com/questions/12421989/networkstream-readasync-with-a-cancellation-token-never-cancels) there are reports for similar issues with NetworkStream.ReadAsync.

I see some ppl report that calling Socket.Close() can interrupt blocked sockets/threads. I am planning to register cancellation token callback and when it is cancelled, if after 10sec timeout the imapclient is still in Idle to call ImapClient.Disconnect which I suppose will call NetworkStream.Dispose() which will call Socket.Close(); I am not planing to use sync version ImapClient.Idle. Do you see any problem with that?

**Comment by jstedfast** [comment]
FWIW, MailKit does not use System.Net.NetworkStream, it uses a custom NetworkStream designed to improve cancellation.

If you think that Socket.Close() would work better than the Socket.DIsconnect(false) call that my previous patch added, I'm willing to change the code to call Close() instead of Disconnect().

Close() *is* what the .NET Core networking team told me to use, after all, they just didn't respond to my question until after I had already changed the code to use Disconnect() and you replied back saying you'd give it a test, so I just figured I'd wait and see how things turned out.

**Comment by ekalchev** [comment]
We are no longer observing issues since we changed to IdleAsync().GetAwaiter().GetResult(). However I still see blocked threads by Socket.Receive. It looks like that is the next network call after IdleAsync cancellation (usually sync call  to ImapClient.Connect) is blocking without throwing any exceptions and cannot be cancelled.

I want to avoid the use IdleAsync().GetAwaiter().GetResult() since our design is to have single mailkit connection in a single thread and with IdleAsync I can see Disconnect event (and probably more stuff) are raised in worker threads. Our app is back to sync version of Idle so we no longer testing this, however I created [test app](https://github.com/ekalchev/ImapIdleTest) and I am trying to reproduce the issue on the machines that we know that the issue occurs.
